app/models/reviewable.rb

```diff
@@ -534,22 +535,60 @@ def self.list_for(
     result
   end
 
+  def self.recent_list_with_pending_first(user)
+    order_clause = DB.sql_fragment(<<~SQL, pending: statuses[:pending])
+      CASE WHEN reviewables.status = :pending THEN 0 ELSE 1 END,
+      CASE WHEN reviewables.status = :pending THEN reviewables.score END DESC,
+      reviewables.created_at DESC
+    SQL
+    query = Reviewable
```

Comment:
Can you check if the resulting query uses any index at all? Looking at the current indexes we have, I have a feeling the query does a seq scan against the whole table.

Reply:
This is the `EXPLAIN ANALYZE` of this query when the user is a group moderator but not admin or full moderator:

```
discourse_development=# EXPLAIN ANALYZE SELECT "reviewables".* FROM "reviewables" WHERE ((reviewables.reviewable_by_moderator AND FALSE) OR (reviewables.reviewable_by_group_id IN (48,10,11,12,0))) AND (reviewables.category_id IS NULL OR reviewables.category_id IN (21,14,11,13,1,29,5,25,22,20,2,26,7,9,18,17,23,17,27)) ORDER BY CASE WHEN reviewables.status = 0 THEN 0 ELSE 1 END, CASE WHEN reviewables.status = 0 THEN reviewables.score END DESC, reviewables.created_at DESC  LIMIT 30;
                                                                           QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=19.62..19.62 rows=1 width=215) (actual time=0.059..0.077 rows=14 loops=1)
   ->  Sort  (cost=19.62..19.62 rows=1 width=215) (actual time=0.058..0.065 rows=14 loops=1)
         Sort Key: (CASE WHEN (status = 0) THEN 0 ELSE 1 END), (CASE WHEN (status = 0) THEN score ELSE NULL::double precision END) DESC, created_at DESC
         Sort Method: quicksort  Memory: 28kB
         ->  Bitmap Heap Scan on reviewables  (cost=8.81..19.61 rows=1 width=215) (actual time=0.027..0.036 rows=14 loops=1)
               Recheck Cond: (reviewable_by_group_id = ANY ('{48,10,11,12,0}'::integer[]))
               Filter: ((category_id IS NULL) OR (category_id = ANY ('{21,14,11,13,1,29,5,25,22,20,2,26,7,9,18,17,23,17,27}'::integer[])))
               Heap Blocks: exact=1
               ->  Bitmap Index Scan on index_reviewables_on_reviewable_by_group_id  (cost=0.00..8.81 rows=9 width=0) (actual time=0.015..0.016 rows=14 loops=1)
                     Index Cond: (reviewable_by_group_id = ANY ('{48,10,11,12,0}'::integer[]))
 Planning Time: 0.094 ms
 Execution Time: 0.136 ms
```

And this is when the user is a full moderator but not admin:

```
discourse_development=# EXPLAIN ANALYZE SELECT "reviewables".* FROM "reviewables" WHERE ((reviewables.reviewable_by_moderator AND TRUE) OR (reviewables.reviewable_by_group_id IN (10,11,12,13,14,46))) AND (reviewables.category_id IS NULL OR reviewables.category_id IN (21,14,11,13,1,29,5,25,22,20,2,26,7,9,18,17,23,4,12,19,24,27,28,30)) ORDER BY CASE WHEN reviewables.status = 0 THEN 0 ELSE 1 END, CASE WHEN reviewables.status = 0 THEN reviewables.score END DESC, reviewables.created_at DESC  LIMIT 30;
                                                                                                                          QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=27.23..27.28 rows=22 width=215) (actual time=0.058..0.081 rows=21 loops=1)
   ->  Sort  (cost=27.23..27.28 rows=22 width=215) (actual time=0.056..0.066 rows=21 loops=1)
         Sort Key: (CASE WHEN (status = 0) THEN 0 ELSE 1 END), (CASE WHEN (status = 0) THEN score ELSE NULL::double precision END) DESC, created_at DESC
         Sort Method: quicksort  Memory: 31kB
         ->  Seq Scan on reviewables  (cost=0.00..26.73 rows=22 width=215) (actual time=0.012..0.025 rows=21 loops=1)
               Filter: ((reviewable_by_moderator OR (reviewable_by_group_id = ANY ('{10,11,12,13,14,46}'::integer[]))) AND ((category_id IS NULL) OR (category_id = ANY ('{21,14,11,13,1,29,5,25,22,20,2,26,7,9,18,17,23,4,12,19,24,27,28,30}'::integer[]))))
 Planning Time: 0.083 ms
 Execution Time: 0.115 ms
```

And finally this is when the user is an admin:

```
discourse_development=# EXPLAIN ANALYZE SELECT "reviewables".* FROM "reviewables" ORDER BY CASE WHEN reviewables.status = 0 THEN 0 ELSE 1 END, CASE WHEN reviewables.status = 0 THEN reviewables.score END DESC, reviewables.created_at DESC  LIMIT 30;
                                                                       QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=25.59..25.66 rows=30 width=215) (actual time=0.068..0.092 rows=21 loops=1)
   ->  Sort  (cost=25.59..26.46 rows=350 width=215) (actual time=0.067..0.076 rows=21 loops=1)
         Sort Key: (CASE WHEN (status = 0) THEN 0 ELSE 1 END), (CASE WHEN (status = 0) THEN score ELSE NULL::double precision END) DESC, created_at DESC
         Sort Method: quicksort  Memory: 31kB
         ->  Seq Scan on reviewables  (cost=0.00..15.25 rows=350 width=215) (actual time=0.014..0.025 rows=21 loops=1)
 Planning Time: 0.090 ms
 Execution Time: 0.119 ms
```

I think none of them does a full table scan? I got the query by running `puts Reviewable.recent_list_with_pending_first(User.find(1)).to_sql.gsub("\n", " ")` and then ran `EXPLAIN ANALYZE` in a `psql` session.

Reply:
Oh wait there's `Seq Scan on reviewables`. That's not good right?

Reply:
> Oh wait there's Seq Scan on reviewables. That's not good right?

Yea the performance of query will degrade as the number of reviewables increases.

Reply:
I refactored this code to split this DB query into 2 queries - first one loads pending records, and the second one loads additional records if the first one returns fewer records than the limit. As far as I can tell it's not possible to make the planner use an index for the original query because it has multiple `CASE WHEN` in its `ORDER BY` clause. I've also added a multicolumn index for `(score DESC, created_at DESC)` to avoid full table scans in the 2 queries.